src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs

```diff
@@ -33,10 +33,18 @@ public sealed partial class JsonObject : JsonNode
         /// <param name="options">Options to control the behavior.</param>
         public JsonObject(IEnumerable<KeyValuePair<string, JsonNode?>> properties, JsonNodeOptions? options = null) : this(options)
         {
+            JsonPropertyDictionary<JsonNode?> dictionary =
+                properties is ICollection<KeyValuePair<string, JsonNode?>> propertiesCollection
+                    ? new JsonPropertyDictionary<JsonNode?>(IsCaseInsensitive, propertiesCollection.Count)
+                    : new JsonPropertyDictionary<JsonNode?>(IsCaseInsensitive);
+
             foreach (KeyValuePair<string, JsonNode?> node in properties)
             {
-                Add(node.Key, node.Value);
+                dictionary.Add(node.Key, node.Value);
+                node.Value?.AssignParent(this); // TODO: how to avoid copying this line from the `Add` existing method??
```

Comment:
I don't think DRY is a huge concern here, it's not a lot of code this is repeating. The benefits of not checking if `_dictionary` is populated on every `Add` operation is substantial.

Reply:
Speaking of perf, is there a performance-related reason why the internal `Dictionary` property is defined like:

```csharp
internal JsonPropertyDictionary<JsonNode?> Dictionary => _dictionary is { } dictionary ? dictionary : InitializeDictionary();
```

instead of using the null-coalescing operator?
```csharp
internal JsonPropertyDictionary<JsonNode?> Dictionary => _dictionary ?? InitializeDictionary();
```

Reply:
https://github.com/dotnet/runtime/pull/96486/commits/d9d68236cf6940805381814c6176f16a33ff660c and https://github.com/dotnet/runtime/pull/96486/commits/b01922f96cd6ed751a1b63fa08b0f9b9a4fad461